File: src/conductances/SK_E2.py
"""
SK_E2: Calcium-activated potassium conductance

Extracted from BBP NEURON mechanism: SK_E2.mod
Neuron type: pyramidal neuron

State variables:
    z: calcium-driven activation gate
"""

Z_TAU = 1.0  # ms
CA_FLOOR = 1e-7  # mM

# euler's constant to 30 digits of precision
e = 2.718281828459045235360287471352

# maximal peak conductance density of channel per unit membrane area (mho/cm^2)
gSK_E2bar = 0.000001

def derivatives(z, ca):
    """
    Compute steady-state value,
    then compute time derivative.

    Inputs:
        - z (float): calcium-driven activation gate
        - ca (float): intracellular calcium concentration (mM)

    Returns: 
        - dz_dt (float): time derivative of z
    """

    if ca < CA_FLOOR:
        ca = ca + CA_FLOOR

    zInf = 1.0 / (1.0 + (0.00043 / ca)**4.8)

    dz_dt = (zInf - z) / Z_TAU
    return dz_dt

def current(z, V, E_K, dt=0.025):
    """
    Compute current based on gating variables and membrane potential.

    Inputs:
        - z (float): calcium-driven activation gate
        - V (float): membrane potential (mV)
        - E_K (float): calcium reversal potential (mV)
        - time step (float): time step of derivative measurement

    Returns: 
        - I_K (float): current (mA/cm^2)
    """
    if ca < CA_FLOOR:
        ca = ca + CA_FLOOR

    zInf = 1.0 / (1.0 + (0.00043 / ca)**4.8)
    
    # cnexp update
    z = zInf + (z - zInf) * e**(-dt/Z_TAU)
    
    gSK_E2 = gSK_E2bar * z
    I_K = gSK_E2 * (E_K)
    return I_K

File: src/cpp/SK_E2.cpp
#include "SK_E2.hpp"

#include <cmath>

namespace {
const scalar_t e = 2.718281828459045235360287471352;
const scalar_t Z_TAU = 1.0;
const scalar_t CA_FLOOR = 1e-7;
const scalar_t gSK_E2bar = 0.000001;
}

scalar_t SK_E2_derivatives(scalar_t z, scalar_t ca) {
    if (ca < CA_FLOOR) {
        ca = ca + CA_FLOOR;
    }

    const scalar_t zInf = 1.0 / (1.0 + std::pow((0.00043 / ca), 4.8));
    return (zInf - z) / Z_TAU;
}

scalar_t SK_E2_current(scalar_t z, scalar_t ca, scalar_t V, scalar_t E_K, scalar_t dt) {
    if (ca < CA_FLOOR) {
        ca = ca + CA_FLOOR;
    }

    const scalar_t zInf = 1.0 / (1.0 + std::pow((0.00043 / ca), 4.8));

    z = zInf + (z - zInf) * std::pow(e, (-dt / Z_TAU));

    const scalar_t gSK_E2 = gSK_E2bar * z;
    const scalar_t I_K = gSK_E2 * E_K;
    return I_K;
}